### PR TITLE
feature: Specify BigQuery dataset for staging data for batch_serve_to_df

### LINKED ISSUE #1622: Specify BigQuery dataset for staging data for batch_serve_to_df
**Is your feature request related to a problem? Please describe.**
We would like to use Vertex Feature Store DataFrame API ([batch_serve_to_df](https://github.com/googleapis/python-aiplatform/blob/3224ae3402e9493866dd4958d011a431968b9c2c/google/cloud/aiplatform/featurestore/featurestore.py#L1087-L1281)) to integrate VFS with Python-based workflows (training and batch inference) that are already standardized on Pandas DataFrame. However, because this function creates an ephemeral GBQ dataset, it requires `bigquery.datasets.create` permission; we have use cases where users will not have this permission. (Instead, individuals typically follow other managed workflows to provision GBQ datasets, not amenable to this function.)

 **Describe the solution you'd like**
Provide optional GBQ dataset parameter. If set, provision ephemeral tables (with randomized names to avoid collision during concurrent runs) within specified dataset. (Otherwise, current logic is unimpacted.)

**Describe alternatives you've considered**
- Adopt other output methods in AIP Vertex Feature Store client (e.g., `batch_serve_to_bq`, `batch_serve_to_gcs`) and then manually load into DataFrame. Problems: requires a lot of boilerplate (e.g., massaging timestamps, I/O, clean-up), and error prone (e.g., massaging timestamps, resource cleanup).
- Create reusable utility methods that uses other output methods in AIP Vertex Feature Store client (e.g., `batch_serve_to_bq`, `batch_serve_to_gcs`) and manually load into DataFrame. Problems: duplicative of `batch_serve_to_df`, fragmented experience (requires internal SDKs and documentation), resource cleanup.

### PR DESCRIPTION
**Pull request by bryanesmith**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-aiplatform/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #1622 🦕

### CODE REVIEW

**Comment by rosiezou** [comment]
@sararob do you have bandwidth to review this PR?

**Comment by sararob** [comment]
We are doing maintenance on our repo starting today and need to close all open PRs. When this is done next week, you'll be able to open a new PR.

**Comment by bryanesmith** [comment]
Hi @rosiezou, I'm unclear on the path to either merging or closing out the PR. If you have any guidance, I'd appreciate it!

**Comment by rosiezou** [comment]
Hi @bryanesmith. The PR looks good from an SDK perspective. Once I check with our cross-functional teammates and make sure that the additional function behavior is consistent with the underlying service behavior, we can go ahead and merge this PR. I'm hoping to have this PR approved by end of this week.

**Review by rosiezou — APPROVED** [review]
Spoke with cross-functional teammates & confirmed that this is expected behavior. LGTM.

File: google/cloud/aiplatform/featurestore/featurestore.py [hunk 1]
```diff
@@ -1273,9 +1299,39 @@ def batch_serve_to_df(
...
+            # clean up: if user specified BigQuery dataset, delete ephemeral tables
+            else:
+                bigquery_client.delete_table(temp_bq_batch_serve_table_id)
+                bigquery_client.delete_table(temp_bq_read_instances_table_id)
 
         return pd.concat(frames, ignore_index=True) if frames else pd.DataFrame(frames)
+
+    def _get_ephemeral_bq_full_dataset_id(
+        self, featurestore_id: str, project_number: str
+    ) -> str:
```
Comment (rosiezou) [inline google/cloud/aiplatform/featurestore/featurestore.py:1318]: @bryanesmith can you add doc strings to this method?

File: google/cloud/aiplatform/featurestore/featurestore.py [hunk 2]
```diff
@@ -1273,9 +1299,39 @@ def batch_serve_to_df(
...
+        project_id = resource_manager_utils.get_project_id(
+            project_number=project_number,
+            credentials=self.credentials,
+        )
+
+        return f"{project_id}.{temp_bq_dataset_name}"[:1024]
+
+    def _create_ephemeral_bq_dataset(
+        self, bigquery_client: bigquery.Client, dataset_id: str
+    ) -> "bigquery.Dataset":
```
Comment (rosiezou) [inline google/cloud/aiplatform/featurestore/featurestore.py:1332]: please add doc strings to this method as well

**Review by rosiezou — CHANGES_REQUESTED** [review]

File: google/cloud/aiplatform/featurestore/featurestore.py (hunk 2 above)
Comment (bryanesmith) [reply google/cloud/aiplatform/featurestore/featurestore.py:1343]: fixed in 7020556

File: google/cloud/aiplatform/featurestore/featurestore.py (hunk 1 above)
Comment (bryanesmith) [reply google/cloud/aiplatform/featurestore/featurestore.py:1318]: fixed in 7020556

File: google/cloud/aiplatform/featurestore/featurestore.py [hunk 3]
```diff
@@ -1210,34 +1216,54 @@ def batch_serve_to_df(
...
         temp_bq_read_instances_table_id = (
-            f"{temp_bq_dataset_id}.{temp_bq_read_instances_table_name}"
+            f"{temp_bq_full_dataset_id}.{temp_bq_read_instances_table_name}"
         )
 
         try:
 
+            # partial config to ensure the timestamp column has Timestamp and
+            #   not Datetime
+            job_config = bigquery.LoadJobConfig(
```
Comment (sasha-gitg) [inline google/cloud/aiplatform/featurestore/featurestore.py:1255]: This seems out-of-scope for the proposed change. Why was this included?

Comment (bryanesmith) [reply google/cloud/aiplatform/featurestore/featurestore.py:1255]: busted; I'll pull that out and create separate issue/pr :)

Comment (bryanesmith) [reply google/cloud/aiplatform/featurestore/featurestore.py]: removed this in [a8f052e](https://github.com/googleapis/python-aiplatform/pull/1623/commits/a8f052e734d0b68fc8d85e852df13a2f1417078f)

**Comment by diemtvu** [comment]
lgtm

**Review by rosiezou — APPROVED** [review]

**Review by junkourata — APPROVED** [review]

**Review by sasha-gitg — APPROVED** [review]
Thanks Bryan!

**Review by sasha-gitg — APPROVED** [review]

**Review by sasha-gitg — APPROVED** [review]

**Comment by sararob** [comment]
Thank you for this contribution! These changes have been committed to `main`, so I'm closing the PR.
